simulation/simulation.py

```diff
@@ -0,0 +1,400 @@
+                os.makedirs(test_folder)
+        else:
+            raise ValueError("Unknown task type: %s" % type(task))
+        if self._profile:
+            cProfile.runctx(statement="self._simulate()",
+                            locals={"self": self},
+                            globals={},
+                            filename="../stats")
+        else:
+            self._simulate()
+
+    def _simulate(self):
+        if self._profile:
+            logging.info("Profiling mode enabled, running first partition in single thread")
+            self._simulate_partition(0)
+        elif self._max_cores == 1:
+            # Run single thread in main thread for easier debugging:
+            for partition_i in range(self._task.partition_count):
+                self._simulate_partition(partition_i)
+        else:
+            pool = multiprocessing.get_context("spawn").Pool(processes=self._max_cores)
+            tasks = range(self._task.partition_count)
+            work = self._simulate_partition
+            for _ in tqdm.tqdm(pool.imap_unordered(work, tasks), total=len(tasks)):
+                pass
+            pool.close()
+        logging.info("Finished simulation")
+
+    def _simulate_partition(self, partition_i: int):
+        # This function is executed within a thread
+        # Need to re-configure logger because we're in a thread:
+        self._configure_logger()
+        logging.debug("Starting partition %s of %s", partition_i, self._task.partition_count)
+        if isinstance(self._task, PreTrainingTask):
+            self._simulate_person_set(partition_i=partition_i,
+                                      persons_per_partition=self._task.person_count // self._task.partition_count,
+                                      output_folder=os.path.join(self._root_folder, PRETRAINING_FOLDER))
+        elif isinstance(self._task, PredictionTask):
+            self._simulate_person_set(partition_i=partition_i,
+                                      persons_per_partition=self._task.train_person_count // self._task.partition_count,
+                                      output_folder=os.path.join(self._root_folder, TRAIN_FOLDER))
+            self._simulate_person_set(partition_i=partition_i,
+                                      persons_per_partition=self._task.test_person_count // self._task.partition_count,
+                                      output_folder=os.path.join(self._root_folder, TEST_FOLDER))
+        else:
+            raise ValueError("Unknown task type: %s" % type(self._task))
+        logging.debug("Finished partition %s of %s", partition_i, self._task.partition_count)
+
+    def _simulate_person_set(self,
+                             partition_i: int,
+                             persons_per_partition: int,
+                             output_folder: str):
+        if isinstance(self._task, PredictionTask):
+            self._cdm_data = cdm_data.CdmDataWithLabels()
```

Comment:
A quick question about how `multiprocessing.Pool` works with the class method `_simulate_partition`. After creating multiple processes, each process will call `_simulate_person_set`, and re-instantiate `self._cdm_data`. I wonder if multiple processes will compete for the class member `self._cdm_data` and cause data inconsistency. Is the `self` object actually getting cloned to each individual process?

Reply:
Yes, Pool creates a clone, so everything set before Pool will be available to all threads. Everything set in a thread stays in the thread, unless you explicitly create shared memory objects.